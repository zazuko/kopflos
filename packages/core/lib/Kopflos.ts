import type { IncomingHttpHeaders, OutgoingHttpHeaders } from 'node:http'
import type { DatasetCore, NamedNode, Stream } from '@rdfjs/types'
import type { GraphPointer, MultiPointer } from 'clownface'
import type { Options as EndpointOptions, StreamClient } from 'sparql-http-client/StreamClient.js'
import type { ParsingClient } from 'sparql-http-client/ParsingClient.js'
import { CONSTRUCT } from '@tpluscode/sparql-builder'
import { IN } from '@tpluscode/sparql-builder/expressions'
import type { KopflosEnvironment } from './env/index.js'
import { createEnv } from './env/index.js'
import type { ResourceShapeLookup, ResourceShapeMatch } from './resourceShape.js'
import defaultResourceShapeLookup from './resourceShape.js'
import { responseOr } from './responseOr.js'
import type { ResourceLoader, ResourceLoaderLookup } from './resourceLoader.js'
import { fromOwnGraph, findResourceLoader } from './resourceLoader.js'
import type { Handler, HandlerArgs, HandlerLookup } from './handler.js'
import { loadHandler } from './handler.js'
import type { HttpMethod } from './httpMethods.js'

interface KopflosRequest {
  iri: NamedNode
  method: HttpMethod
  headers: IncomingHttpHeaders
}

type ResultBody = Stream | DatasetCore | GraphPointer | Error
export interface ResultEnvelope {
  body?: ResultBody | string
  status?: number
  headers?: OutgoingHttpHeaders
}
export type KopflosResponse = ResultBody | ResultEnvelope

export interface Kopflos {
  get env(): KopflosEnvironment
  get apis(): MultiPointer
  handleRequest(req: KopflosRequest): Promise<ResultEnvelope>
}

interface Clients {
  stream: StreamClient
  parsed: ParsingClient
}

type Endpoint = string | EndpointOptions | Clients

export interface KopflosConfig {
  sparql: Record<string, Endpoint> & { default: Endpoint }
  codeBase?: string
}

export interface Options {
  dataset?: DatasetCore
  resourceShapeLookup?: ResourceShapeLookup
  resourceLoaderLookup?: ResourceLoaderLookup
  handlerLookup?: HandlerLookup
}

type Dataset = ReturnType<KopflosEnvironment['dataset']>

export default class Impl implements Kopflos {
  readonly dataset: Dataset
  readonly env: KopflosEnvironment

  constructor(private readonly config: KopflosConfig, private readonly options: Options = {}) {
    this.env = createEnv(config)

    this.dataset = this.env.dataset([
      ...options.dataset || [],
    ])
  }

  get graph() {
    return this.env.clownface({ dataset: this.dataset })
  }

  get apis(): MultiPointer {
    return this.graph.has(this.env.ns.rdf.type, this.env.ns.kopflos.Api)
  }

  async handleRequest(req: KopflosRequest): Promise<ResultEnvelope> {
    const result = await responseOr(this.findResourceShape(req.iri), (resourceShapeMatch: ResourceShapeMatch) => {
      const resourceShape = this.graph.node(resourceShapeMatch.resourceShape)

      return responseOr(this.findResourceLoader(resourceShape), loader => {
        const coreRepresentation = loader(req.iri, this)

        return responseOr(this.loadHandler(req.method, resourceShapeMatch, coreRepresentation), async handler => {
          const resourceGraph = this.env.clownface({
            dataset: await this.env.dataset().import(coreRepresentation),
          })
          const args: HandlerArgs = {
            resourceShape,
            env: this.env,
            subject: resourceGraph.node(resourceShapeMatch.subject),
            property: undefined,
            object: undefined,
          }
          if ('property' in resourceShapeMatch) {
            args.property = resourceShapeMatch.property
            args.object = resourceGraph.node(resourceShapeMatch.object)
          }

          return handler(args)
        })
      })
    })

    if (this.isEnvelope(result)) {
      return result
    }

    return {
      status: 200,
      body: result,
    }
  }

  async findResourceShape(iri: NamedNode): Promise<ResourceShapeMatch | KopflosResponse> {
    const resourceShapeLookup = this.options.resourceShapeLookup || defaultResourceShapeLookup

    return responseOr(resourceShapeLookup(iri, this), async candidates => {
      if (candidates.length === 0) {
        return { status: 404 }
      }

      if (candidates.length > 1) {
        return new Error('Multiple resource shapes found')
      }

      return candidates[0]
    })
  }

  async findResourceLoader(resourceShape: GraphPointer): Promise<ResourceLoader | KopflosResponse> {
    const resourceLoaderLookup = this.options.resourceLoaderLookup || findResourceLoader

    const loader = await resourceLoaderLookup(resourceShape, this.env)

    return loader || fromOwnGraph
  }

  async loadHandler(method: HttpMethod, resourceShapeMatch: ResourceShapeMatch, coreRepresentation: Stream): Promise<Handler | KopflosResponse> {
    const handlerLookup = this.options.handlerLookup || loadHandler

    const handler = await handlerLookup(resourceShapeMatch, method, this)

    if (handler) {
      return handler
    }

    if (!('property' in resourceShapeMatch) && (method === 'GET' || method === 'HEAD')) {
      return {
        status: 200,
        body: coreRepresentation,
      }
    }

    return { status: 405 }
  }

  private isEnvelope(arg: KopflosResponse): arg is ResultEnvelope {
    return 'body' in arg || 'status' in arg
  }

  static async fromGraphs(kopflos: Impl, ...graphs: Array<NamedNode | string>): Promise<void> {
    const graphsIris = graphs.map(graph => typeof graph === 'string' ? kopflos.env.namedNode(graph) : graph)
    const quads = CONSTRUCT`?s ?p ?o `
      .WHERE`
        GRAPH ?g {
          ?s ?p ?o
        }
        
        FILTER (?g ${IN(...graphsIris)})
      `.execute(kopflos.env.sparql.default.stream)

    for await (const quad of quads) {
      kopflos.dataset.add(quad)
    }
  }
}
