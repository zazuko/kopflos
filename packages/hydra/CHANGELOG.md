# @kopflos-cms/hydra

## 0.1.5

### Patch Changes

- bfc7e90: `hydra:Collection`: `sh:shapesGraph` can be used to import shapes used in `kl-hydra:memberShape` and `kl-hydra:memberCreateShape`
- bfc7e90: Update `@kopflos-cms/shacl` to use `sh:shapesGraph`
- Updated dependencies [bfc7e90]
  - @kopflos-cms/core@0.5.2

## 0.1.4

### Patch Changes

- 21b4319: Assign partial collection view URI
- Updated dependencies [bab5fc1]
- Updated dependencies [e1c91d1]
  - @kopflos-cms/core@0.5.1

## 0.1.3

### Patch Changes

- Updated dependencies [3f737fd]
  - @kopflos-cms/core@0.5.0

## 0.1.2

### Patch Changes

- 6328d5c: Collections: support `kl-hydra:memberCreateShape` and `kl-hydra:memberQueryShape` which can be used instead of `kl-hydra:memberShape` to separate querying from validation

## 0.1.1

### Patch Changes

- a95d3f7: Some files missing from package

## 0.1.0

### Minor Changes

- dcddc48: Support for sending `POST` requests to a `hydra:Collection`
- 67c5db0: Configurable SPARQL Endpoint to use for fetching collection members
- b84bd82: Created extensible method for collection paging strategies
- be60b63: Support for `hydra:Collection` resource shapes

### Patch Changes

- Updated dependencies [ad51eef]
- Updated dependencies [b84bd82]
- Updated dependencies [b84bd82]
- Updated dependencies [be60b63]
- Updated dependencies [67c5db0]
- Updated dependencies [608724b]
  - @kopflos-cms/core@0.4.0
