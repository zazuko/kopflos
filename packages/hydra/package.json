{
  "name": "@kopflos-cms/hydra",
  "version": "0.0.0",
  "main": "index.js",
  "license": "MIT",
  "type": "module",
  "scripts": {
    "test": "mocha",
    "build": "tsc",
    "prepack": "npm run build"
  },
  "exports": {
    ".": "./index.js"
  },
  "files": [
    "CHANGELOG.md",
    "*.js",
    "*.d.ts",
    "handlers/*.js",
    "handlers/*.d.ts"
  ],
  "dependencies": {
    "@hydrofoil/shape-to-query": "^0.13.6",
    "@kopflos-cms/core": "^0.3.1",
    "@rdfine/hydra": "^0.10.5",
    "@sindresorhus/merge-streams": "^4.0.0",
    "@tpluscode/rdf-ns-builders": "^4.3.0",
    "@tpluscode/rdfine": "^0.7.10",
    "@zazuko/env": "^2.5.1",
    "http-errors": "^2.0.0",
    "is-graph-pointer": "^2.1.0",
    "rdf-literal": "^1.3.2"
  },
  "devDependencies": {
    "@rdfjs-elements/formats-pretty": "^0.6.8",
    "@types/http-errors": "^2.0.4",
    "@zazuko/env-node": "^2.1.4",
    "chai": "^5.1.2",
    "mocha-chai-rdf": "^0.1.6"
  },
  "mocha": {
    "extension": [
      "ts"
    ],
    "spec": "test/**/*.test.ts",
    "loader": "ts-node/esm/transpile-only",
    "require": "../../mocha-setup.js"
  }
}
