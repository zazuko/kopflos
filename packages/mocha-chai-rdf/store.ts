/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
