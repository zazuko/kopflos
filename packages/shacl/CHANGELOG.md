# @kopflos-cms/shacl

## 0.1.1

### Patch Changes

- a95d3f7: Some files missing from package

## 0.1.0

### Minor Changes

- bd6aa25: SHACL Validation decorator
