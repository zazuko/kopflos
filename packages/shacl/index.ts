import type { HandlerArgs, Kopflos, KopflosPlugin, KopflosPluginConstructor } from '@kopflos-cms/core'
import type { DatasetCore } from '@rdfjs/types'

type ExtendingTerms = 'shacl#shapeSelector'

declare module '@kopflos-cms/core/ns.js' {
  interface KopflosTerms extends Record<ExtendingTerms, never> {
  }
}

interface AdditionalDataGraphTriplesArgs extends HandlerArgs{
  shapesGraph: DatasetCore
}

export interface Options {
  loadDataGraph?(arg: AdditionalDataGraphTriplesArgs): Promise<DatasetCore>
}

export interface ShaclPlugin extends KopflosPlugin {
  readonly options: Options
}

declare module '@kopflos-cms/core' {
  interface PluginConfig {
    '@kopflos-cms/shacl'?: Options
  }

  interface Plugins {
    '@kopflos-cms/shacl': ShaclPlugin
  }
}

const decoratorModule = new URL('./lib/decorator.js#default', import.meta.url).toString()

export default function factory(options: Options = {}): KopflosPluginConstructor<ShaclPlugin> {
  return class implements ShaclPlugin {
    public readonly options = Object.freeze(options)

    public readonly name = '@kopflos-cms/shacl'

    constructor(private readonly instance: Kopflos) {
    }

    async apiTriples() {
      const { env } = this.instance

      const apis = env.clownface()
        .node(this.instance.apis)

      const impl = apis.blankNode()
        .addOut(env.ns.rdf.type, env.ns.code.EcmaScriptModule)
        .addOut(env.ns.code.link, env.namedNode(decoratorModule))

      apis
        .addOut(env.ns.kl.decorator, decorator => {
          decorator.addOut(env.ns.code.implementedBy, impl)
        })

      return apis.dataset
    }
  }
}
