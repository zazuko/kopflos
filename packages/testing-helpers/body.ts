import type { Readable } from 'node:stream'
import type { NamedNode } from '@rdfjs/types'
import type { Body } from '@kopflos-cms/core'
import rdf from '@zazuko/env-node'
import type { Dataset } from '@zazuko/env/lib/DatasetExt.js'

declare module '@rdfjs/types' {
  interface Stream extends Readable {}
}

export function asBody(dataset: Dataset, term: NamedNode): Body {
  return {
    isRDF: true,
    get quadStream() {
      return dataset.toStream()
    },
    get dataset() {
      return dataset
    },
    async pointer() {
      return rdf.clownface({ dataset, term })
    },
  } as unknown as Body
}
