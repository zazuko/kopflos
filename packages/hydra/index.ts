import type { Kopflos, KopflosEnvironment, KopflosPlugin, KopflosPluginConstructor } from '@kopflos-cms/core'
import type { NamedNode } from '@rdfjs/types'
import type { DerivedEnvironment } from '@zazuko/env'
import E from '@zazuko/env/Environment.js'
import { RdfineFactory } from '@tpluscode/rdfine'
import { HydraFactory } from '@rdfine/hydra/Factory'
import type { Environment } from '@rdfjs/environment/Environment.js'
import { createDefaultShapes, createHandlers } from './lib/resourceShapes.js'
import type { PartialCollectionStrategy } from './lib/partialCollection/index.js'
import limitOffsetStrategy from './lib/partialCollection/limitOffsetStrategy.js'
import pageIndexStrategy from './lib/partialCollection/pageIndexStrategy.js'

type ExtendingTerms = 'hydra#memberShape'
| 'hydra#memberCreateShape'
| 'hydra#memberQueryShape'
| 'hydra#MemberAssertionConstraintComponent'
| 'hydra'
| 'hydra#DefaultCollectionShape'
| 'hydra#memberUriTemplate'

declare module '@kopflos-cms/core/ns.js' {
  interface KopflosTerms extends Record<ExtendingTerms, never> {
  }
}

export interface Options {
  /**
   * The IRI of the API that the Hydra API Documentation will be generated and served for
   */
  apis: Array<NamedNode | string>
  partialCollectionStrategies?: PartialCollectionStrategy[]
}

declare module '@kopflos-cms/core' {
  interface PluginConfig {
    '@kopflos-cms/hydra'?: Options
  }
}

export interface HydraPlugin extends KopflosPlugin {
  readonly env: DerivedEnvironment<Environment<RdfineFactory | HydraFactory>, KopflosEnvironment>
  readonly partialCollectionStrategies: PartialCollectionStrategy[]
}

declare module '@kopflos-cms/core' {
  interface Plugins {
    hydra: HydraPlugin
  }
}

export default (options : Options): KopflosPluginConstructor => {
  return class implements HydraPlugin {
    readonly name = 'hydra'

    readonly env: DerivedEnvironment<Environment<RdfineFactory | HydraFactory>, KopflosEnvironment>
    readonly partialCollectionStrategies: PartialCollectionStrategy[]

    constructor(private readonly instance: Kopflos) {
      this.env = new E([RdfineFactory, HydraFactory], { parent: instance.env })
      this.partialCollectionStrategies = [
        ...options.partialCollectionStrategies ?? [],
        limitOffsetStrategy,
        pageIndexStrategy,
      ]
    }

    async onStart() {
      const { kopflos: kl } = this.env.ns

      const dataset = createDefaultShapes(this.env, options)

      await this.env.sparql.default.stream.store.put(dataset.toStream(), {
        graph: kl.hydra,
      })
    }

    async apiTriples() {
      return createHandlers(this.instance)
    }
  }
}
