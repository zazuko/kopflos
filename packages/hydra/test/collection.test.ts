import * as fs from 'node:fs'
import type { Body, KopflosConfig, ResultEnvelope } from '@kopflos-cms/core'
import Kopflos from '@kopflos-cms/core'
import { createStore } from 'mocha-chai-rdf/store.js'
import $rdf from '@zazuko/env-node'
import { expect } from 'chai'
import type { NamedNode, Stream } from '@rdfjs/types'
import * as ns from '@tpluscode/rdf-ns-builders'
import { toRdf } from 'rdf-literal'
import shacl from '@kopflos-cms/shacl'
import type { Dataset } from '@zazuko/env/lib/DatasetExt.js'
import inMemoryClients from '../../testing-helpers/in-memory-clients.js'
import hydra from '../index.js'
import { asBody } from '../../testing-helpers/body.js'

const baseIri = 'http://example.org'
const ex = $rdf.namespace(baseIri + '/')

describe('@kopflos-cms/hydra', () => {
  beforeEach(createStore(import.meta.url, {
    format: 'trig',
    loadAll: true,
  }))

  let config: KopflosConfig
  let clients: ReturnType<typeof inMemoryClients>
  beforeEach(function () {
    this.rdf.store.load(fs.readFileSync(new URL('./data/municipalities.ttl', import.meta.url)).toString(), {
      format: 'text/turtle',
    })
    this.rdf.store.load(fs.readFileSync(new URL('./data/districts.ttl', import.meta.url)).toString(), {
      format: 'text/turtle',
    })

    clients = inMemoryClients(this.rdf)
    config = {
      baseIri,
      sparql: {
        default: clients,
        lindas: 'https://lindas.admin.ch/query',
      },
    }
  })

  async function startKopflos({ apis = [ex('readonly-api')] } = {}) {
    const kopflos = new Kopflos({
      ...config,
      apiGraphs: apis.map(api => api.value),
    }, {
      plugins: [hydra({
        apis,
      }), shacl()],
    })
    await kopflos.start()
    await kopflos.loadApiGraphs()

    return kopflos
  }

  describe('hydra:Collection', () => {
    context('get', () => {
      context('when collection has no memberAssertion', () => {
        it('should return 500', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: ex['municipalities/no-assertions'],
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          expect(res.status).to.equal(500)
        })
      })

      context('when collection is not readable', () => {
        it('should return 405', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: ex['municipalities/readable-false'],
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          expect(res.status).to.equal(405)
        })
      })

      context('when collection has single memberAssertion', () => {
        it('should return a stream of members', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const collection = ex['municipalities/all']
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: collection,
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          const dataset = await $rdf.dataset().import(res.body as Stream)
          const pointer = $rdf.clownface({ dataset }).node(collection)
          expect(pointer.out(ns.hydra.member).terms).to.have.length(3449)
          expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
        })
      })

      context('collection paged', () => {
        let res: ResultEnvelope

        context('with limit/offset', () => {
          const collection = ex['municipalities/limit-offset']

          context('first page', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  limit: '10',
                  offset: '0',
                },
                body: {} as Body,
              })
            })

            it('does not have link to previous page', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.previous).term).to.be.undefined
            })
          })

          context('last page', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  limit: '10',
                  offset: '3440',
                },
                body: {} as Body,
              })
            })

            it('does not have link to next page', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.next).term).to.be.undefined
            })
          })

          context('limit and offset given', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  limit: '10',
                  offset: '40',
                },
                body: {} as Body,
              })
            })

            it('returns subset of results', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.member).terms).to.have.length(10)
              expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
            })

            it('includes next page link', async () => {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.next).term)
                .to.deep.eq(ex['municipalities/limit-offset?limit=10&offset=50'])
            })

            it('includes previous page link', async () => {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.previous).term)
                .to.deep.eq(ex['municipalities/limit-offset?limit=10&offset=30'])
            })

            it('includes first page link', async () => {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.first).term)
                .to.deep.eq(ex['municipalities/limit-offset?limit=10&offset=0'])
            })

            it('includes last page link', async () => {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.view).out(ns.hydra.last).term)
                .to.deep.eq(ex['municipalities/limit-offset?limit=10&offset=3440'])
            })
          })

          context('only limit given', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  limit: '10',
                },
                body: {} as Body,
              })
            })

            it('returns subset of results', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.member).terms).to.have.length(10)
              expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
            })
          })

          context('only offset given', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  offset: '10',
                },
                body: {} as Body,
              })
            })

            it('returns subset of results', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.member).terms).to.have.length(10)
              expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
            })
          })
        })

        context('with page index', () => {
          const collection = ex['municipalities/paged']

          context('page index given', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {
                  page: '10',
                },
                body: {} as Body,
              })
            })

            it('returns subset of results', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.member).terms).to.have.length(20)
              expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
            })
          })

          context('page index not given', () => {
            beforeEach(async function () {
              // given
              const kopflos = await startKopflos()

              // when
              res = await kopflos.handleRequest({
                method: 'GET',
                iri: collection,
                headers: {},
                query: {},
                body: {} as Body,
              })
            })

            it('returns subset of results', async function () {
              // then
              const dataset = await $rdf.dataset().import(res.body as Stream)
              const pointer = $rdf.clownface({ dataset }).node(collection)
              expect(pointer.out(ns.hydra.member).terms).to.have.length(20)
              expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(3449))
            })
          })
        })
      })

      context('when collection has multiple memberAssertion', () => {
        it('should return a stream of members', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const collection = ex['municipalities/from-static-district']
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: collection,
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          const dataset = await $rdf.dataset().import(res.body as Stream)
          const pointer = $rdf.clownface({ dataset }).node(collection)
          expect(pointer.out(ns.hydra.member).terms).to.have.length(14)
          expect(pointer.out(ns.hydra.totalItems).term).to.deep.eq(toRdf(14))
        })
      })

      context('when collection is sourced from another endpoint', function () {
        this.timeout(10000)

        it('should return a stream of members', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const collection = ex['countries/from-lindas']
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: collection,
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          const dataset = await $rdf.dataset().import(res.body as Stream)
          const pointer = $rdf.clownface({ dataset }).node(collection)
          expect(pointer.out(ns.hydra.member).terms).to.have.length.above(200)
        })

        it('should fail when endpoint is not configured', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const collection = ex['countries/wrong-endpoint']
          const res = await kopflos.handleRequest({
            method: 'GET',
            iri: collection,
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          expect(res.status).to.equal(500)
        })
      })
    })

    context('post', function () {
      let loadGraph: (graph: NamedNode) => Dataset

      beforeEach(function () {
        loadGraph = (graph: NamedNode) => {
          const quads = this.rdf.store.match(null, null, null, graph)
            .map(quad => $rdf.quad(quad.subject, quad.predicate, quad.object))

          return $rdf.dataset(quads)
        }
      })

      context('when collection is not writable', () => {
        it('should return 405', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: ex['municipalities/writable-false'],
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          expect(res.status).to.equal(405)
        })
      })

      context('when collection has validation', () => {
        it('creates a new resource', async function () {
          // given
          const kopflos = await startKopflos()
          const collection = ex['municipalities/writable-with-validation']
          const newMember = ex('valid-municipality')
          const dataset = loadGraph(newMember)
          $rdf.clownface({ dataset })
            .node(collection)
            .addOut(ns.hydra.member, newMember)

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: collection,
            headers: {},
            query: {},
            body: asBody(dataset, collection),
          })

          // then
          expect(res.status).to.equal(201)
          expect(res.headers?.Location).to.equal(`${ex('municipality/valid-name').value}`)
          const newMemberDataset = await $rdf.dataset()
            .import(clients.stream.store.get(ex('municipality/valid-name')))
          expect(newMemberDataset).canonical.toMatchSnapshot()
        })

        it('returns 409 when destination graph already exists', async function () {
          // given
          const kopflos = await startKopflos()
          const collection = ex['municipalities/writable-with-validation']
          const newMember = ex('municipality/already-exists')
          const dataset = loadGraph(newMember)
          $rdf.clownface({ dataset })
            .node(collection)
            .addOut(ns.hydra.member, newMember)

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: collection,
            headers: {},
            query: {},
            body: asBody(dataset, collection),
          })

          // then
          expect(res.status).to.equal(409)
        })

        it('should return 400 when body is invalid', async function () {
          // given
          const kopflos = await startKopflos()
          const collection = ex['municipalities/writable-with-validation']
          const newMember = ex('invalid-municipality')
          const dataset = loadGraph(newMember)
          $rdf.clownface({ dataset })
            .node(collection)
            .addOut(ns.hydra.member, newMember)

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: collection,
            headers: {},
            query: {},
            body: asBody(dataset, collection),
          })

          // then
          expect(res.status).to.equal(400)
        })

        it('should return 400 when body does not include hydra:member triple', async function () {
          // given
          const kopflos = await startKopflos()
          const body = $rdf.clownface()
            .node(ex())
            .addOut(ns.rdf.type, ex.Foo)
            .addOut(ns.schema.name, 'Foo')

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: ex['municipalities/writable-with-validation'],
            headers: {},
            query: {},
            body: asBody(body.dataset, ex()),
          })

          // then
          expect(res.status).to.equal(400)
        })

        it('should return 400 when body is empty', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: ex['municipalities/writable-with-validation'],
            headers: {},
            query: {},
            body: asBody($rdf.dataset(), ex()),
          })

          // then
          expect(res.status).to.equal(400)
        })

        it('should return 400 when body is not rdf', async function () {
          // given
          const kopflos = await startKopflos()

          // when
          const res = await kopflos.handleRequest({
            method: 'POST',
            iri: ex['municipalities/writable-with-validation'],
            headers: {},
            query: {},
            body: {} as Body,
          })

          // then
          expect(res.status).to.equal(400)
        })
      })
    })
  })
})
