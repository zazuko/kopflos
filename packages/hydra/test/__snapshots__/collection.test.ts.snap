// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/hydra hydra:Collection post when collection has validation creates a new resource 1`] = `
"<http://example.org/municipality/valid-name> <http://schema.org/name> \\"Valid name\\" <http://example.org/municipality/valid-name> .
<http://example.org/municipality/valid-name> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://schema.ld.admin.ch/Municipality> <http://example.org/municipality/valid-name> .
"
`;

exports[`@kopflos-cms/hydra hydra:Collection post when collection has validation when collection uses :hydraCreateShape creates a new resource 1`] = `
"<http://example.org/municipality/valid-name> <http://schema.org/name> \\"Valid name\\" <http://example.org/municipality/valid-name> .
<http://example.org/municipality/valid-name> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://schema.ld.admin.ch/Municipality> <http://example.org/municipality/valid-name> .
"
`;
